Frameworks/CoreData/NSManagedObjectModel-XMLParsing.mm

```diff
@@ -0,0 +1,288 @@
+                                                           _kNSManagedObjectModelEntityUniquenessConstraintsTag,
+                                                           nullptr];
+    return [_tagSet member:tag];
+}
+
+static Class _classForElementName(NSString* elementName) {
+    static StrongId<NSDictionary<NSString*, Class<_NSCDXMLCoding>>> _mappings = @{
+        _kNSManagedObjectModelTag : [NSManagedObjectModel class],
+        _kNSManagedObjectModelEntityTag : [NSEntityDescription class],
+        _kNSManagedObjectModelEntityAttributeTag : [NSAttributeDescription class],
+        _kNSManagedObjectModelEntityRelationshipTag : [NSRelationshipDescription class],
+    };
+    return [_mappings objectForKey:elementName];
+}
+
+// XML data model parsing is done via recursive descent and state management. Each element
+// we encounter can push another state onto the stack. The state dictates what state transitions
+// are available and what is to be done when the state is exited.
+//
+// Some of these states cascade, and their order should be retained.
+enum _NSModelParsingState : unsigned int {
+    _NSModelParsingStateNone = 0,
+    _NSModelParsingStateGenericObject,
+    _NSModelParsingStateModel,
+    _NSModelParsingStateEntity,
+
+    _NSModelParsingStateUserInfo,
+    _NSModelParsingStateUserInfoEntry,
+
+    _NSModelParsingStateCompoundIndices, // object is an array<array<string>>
+    _NSModelParsingStateCompoundIndex, // object is an array<string>
+    _NSModelParsingStateCompoundIndexEntry, // object is a string
+
+    _NSModelParsingStateUniquenessConstraints, // object is an array<array<string>>
+    _NSModelParsingStateUniquenessConstraint, // object is an array<string>
+    _NSModelParsingStateUniquenessConstraintEntry, // object is a string
+};
+
+struct _NSModelParsingContext {
+    StrongId<NSManagedObjectModel> model;
+    std::list<std::tuple<_NSModelParsingState, StrongId<NSObject>>> states;
+
+    void push_state(_NSModelParsingState st, id object) {
+        states.emplace_front(st, object);
+    }
+    std::tuple<_NSModelParsingState, id> pop() {
+        auto state = states.front();
+        states.pop_front();
+        return state;
+    }
+    std::tuple<_NSModelParsingState, id> peek() {
+        return states.front();
+    }
+    _NSModelParsingState state() {
+        return states.empty() ? _NSModelParsingStateNone : std::get<_NSModelParsingState>(states.front());
+    }
+    size_t count() const {
+        return states.size();
+    }
+};
+
+static void* SKIP = nullptr;
+static void* CONTINUE = (void*)0x01;
+
+#define EMIT_STATE(st, ...)                   \
+    do {                                      \
+        context->push_state(st, __VA_ARGS__); \
+        return CONTINUE;                      \
+    } while (0);
+
+static void* xmlCreateStructure(CFXMLParserRef parser, CFXMLNodeRef node, void* info) {
```

Comment:
This is part of the CFXML contract. `CFXMLCreateStructure` callbacks must return a structure to continue parsing, else the element is skipped. In order, the calls become create -> add -> end. That doesn't work for us, however, because we want to realize all our elements before we add them to the hierarchy.
We return _something_ here and move the work into `endStructure` so that we get final control.

---
In reply to: [70722579](https://github.com/Microsoft/WinObjC/pull/610#discussion_r70722579) [](ancestors = 70722579)